### PR TITLE
Remove many `*`/`mul!` methods for AdjOrTrans of sym/herm/diag/triangular

### PR DESCRIPTION
**Pull request by dkarrasch**
I believe these are relicts from the transition to lazy `Adjoint` and `Transpose` wrappers back in the v0.7/v1.0 times, perhaps artifacts of a mechanical translation of previous methods. Types à la `Adjoint{...,Hermitian}`, however, cannot be constructed from (the recommended) calls of `tranpose` and `adjoint`. They can only be constructed via brute force by calling the view constructors (with capital initial letters) directly. That is, in turn, not recommended, because it avoids the multiple dispatch mechanism built into `adjoint`, which may or may not return an `Adjoint` view, depending on types. The same holds, BTW, for adjoints/transposes of triangular matrices (EDIT: and diagonal matrices). In summary, because these wrapped type combinations should not occur in the wild, I don't expect any "slow fallback" issue to arise, but we should check, of course. IIRC, though, in our benchmark suite we may be testing still for `Transpose/Adjoint(A)`, which should be changed if necessary. Independently from that, a pkgeval run should reveal if packages still relies on these methods.

### CODE REVIEW

**Comment by dkarrasch** [comment]
Tests pass locally, and I think they do because multiplication of some brutally constructed `Adjoint(Hermitian(...))` falls back to `generic_matmul`, which is perhaps hard to detect from pkgeval runs etc. The issue is that the first set of methods causes ambiguities, for whose fix we need another 20+ methods, not counting ambiguities caused in downstream packages, cf. https://github.com/JeffFessler/LinearMapsAA.jl/pull/25.

**Comment by dkarrasch** [comment]
Another option to detect issues in packages is to make all of the currently removed methods throw an error, run pkgeval and see what happens. I'll push some smaller test adjustments next.

**Comment by dkarrasch** [comment]
@nanosoldier `runbenchmarks("linalg", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/41188#issuecomment-859482145) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d7fd770_vs_e35028a/report.md). cc @christopher-dG

**Comment by dkarrasch** [comment]
This is ready for review. @andreasnoack @tkf?

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -306,157 +297,59 @@ function *(D::Diagonal, adjA::Adjoint{<:Any,<:AbstractMatrix})
...
-             A::Adjoint{<:Any,<:Diagonal}, B::Adjoint{<:Any,<:RealHermSymComplexSym},
-             alpha::Number, beta::Number) =
-    C .= (adjoint.(A.parent.diag) .* B) .*ₛ alpha .+ C .*ₛ beta
-@inline mul!(C::AbstractMatrix,
-             A::Transpose{<:Any,<:Diagonal}, B::Transpose{<:Any,<:RealHermSymComplexHerm},
-             alpha::Number, beta::Number) =
-    C .= (transpose.(A.parent.diag) .* B) .*ₛ alpha .+ C .*ₛ beta
+
+function mul!(C::AbstractMatrix, Da::Diagonal, Db::Diagonal, alpha::Number, beta::Number)
+    (size(C) == size(Da) == size(Db)) || throw(DimensionMismatch(""))
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/diagonal.jl:332]: This one could use a better error message.

**Review by andreasnoack — COMMENTED** [review]
Great to get rid of all these unused methods. Once the error message has been expanded then this should be good to go.

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 1 above)
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/diagonal.jl:332]: Yeah, I wanted to make it consistent with other error messages that we throw in such situations, and got distracted on my way finding an appropriate example. The official version is, of course, you passed my attention test, congrats, @andreasnoack 🥇 ! 🤣

**Review by andreasnoack — APPROVED** [review]

**Pull request merged**